addons/sale/models/sale_order.py

```diff
@@ -404,20 +411,23 @@ def _compute_team_id(self):
                     user_id=user_id, domain=[('company_id', 'in', [company_id, False])])
             order.team_id = cached_teams[key]
 
-    @api.depends('order_line.price_total')
+    @api.depends('order_line.price_total', 'invoice_ids')
     def _amount_all(self):
         """
         Compute the total amounts of the SO.
         """
         for order in self:
-            amount_untaxed = amount_tax = 0.0
+            amount_untaxed = amount_tax = amount_invoiced = 0.0
             for line in order.order_line:
                 amount_untaxed += line.price_subtotal
                 amount_tax += line.price_tax
+            if order.invoice_ids and order.invoice_status != 'invoiced':
+                amount_invoiced = sum(order.invoice_ids.filtered(lambda x: x.state == 'posted').mapped('amount_total'))
             order.update({
                 'amount_untaxed': amount_untaxed,
                 'amount_tax': amount_tax,
                 'amount_total': amount_untaxed + amount_tax,
+                'amount_residual': amount_untaxed + amount_tax - amount_invoiced,
             })
```

Comment:
First of all, there is no reason at all to compute this field here, it doesn't share any logic with the other amounts (except the for loop :P)
Secondly, 6 new fields (including one stored) for a 'cosmetic' warning feature is way too much IMHO.
(If we really want the feature, most of it could be simplified to `show_partner_credit_warning: bool` & `show_partner_credit_warning_message: Char/Text`, removing most fields.